src/Bicep.Core/Diagnostics/DiagnosticBuilder.cs

```diff
@@ -1611,6 +1614,123 @@ public ErrorDiagnostic UnknownModuleReferenceScheme(string badScheme, ImmutableA
+                TextSpan,
+                "BCP290",
+                "Expected a parameter or type declaration after the decorator.");
+
+            public ErrorDiagnostic ExpectedParameterOrOutputDeclarationAfterDecorator() => new(
+                TextSpan,
+                "BCP291",
+                "Expected a parameter or output declaration after the decorator.");
+
+            public ErrorDiagnostic ExpectedParameterOutputOrTypeDeclarationAfterDecorator() => new(
+                TextSpan,
+                "BCP292",
+                "Expected a parameter, output, or type declaration after the decorator.");
+
+            public ErrorDiagnostic NonLiteralUnionMember() => new(
+                TextSpan,
+                "BCP293",
+                "All members of a union type declaration must be literal values.");
+
+            public ErrorDiagnostic InvalidTypeUnion() => new(
+                TextSpan,
+                "BCP294",
+                "Type unions must be reduceable to a single ARM type (such as 'string', 'int', or 'bool').");
+
+            public ErrorDiagnostic DecoratorNotPermittedOnLiteralType(string decoratorName) => new(
+                TextSpan,
+                "BCP295",
+                $"The '{decoratorName}' decorator may not be used on targets of a union or literal type. The allowed values for this parameter or type definition will be derived from the union or literal type automatically.");
+
+            public ErrorDiagnostic NonConstantTypeProperty() => new(
+                TextSpan,
+                "BCP296",
+                "Property names on types must be compile-time constant values.");
+
+
+            public ErrorDiagnostic CannotUseFunctionAsTypeDecorator(string functionName) => new(
+                TextSpan,
+                "BCP297",
+                $"Function \"{functionName}\" cannot be used as a type decorator.");
+
+            public ErrorDiagnostic CyclicTypeSelfReference() => new(
+                TextSpan,
+                "BCP298",
+                "This type definition includes itself as required component, which creates a constraint that cannot be fulfilled.");
+
+            public ErrorDiagnostic CyclicType(IEnumerable<string> cycle) => new(
+                TextSpan,
+                "BCP299",
+                $"This type definition includes itself as a required component via a cycle (\"{string.Join("\" -> \"", cycle)}\").");
+
+            public ErrorDiagnostic ExpectedTypeLiteral() => new(
+                TextSpan,
+                "BCP300",
+                $"Expected a type literal at this location. Please specify a concrete value or a reference to a literal type.");
+
+            public ErrorDiagnostic TypeNameMasksAmbientType(string conflictingName) => new(
+                TextSpan,
+                "BCP301",
+                $"A user-defined type may not be named '{conflictingName},' as this masks the ARM type of the same name.");
```

Comment:
`'{conflictingName},'` - comma in the wrong place, and our convention is to use double-quotes vs single-quotes (in fact, to avoid the mix-up with string literal type names!)